pkg/plugin/cgroup.go

```diff
@@ -0,0 +1,271 @@
+// It returns an empty string if the container has no Linux configuration or cgroups path.
+func GetContainerCgroupsV2AbsPath(container *api.Container) string {
+	if container == nil || container.Linux == nil || container.Linux.CgroupsPath == "" {
+		return ""
+	}
+
+	cgroupPath := container.Linux.CgroupsPath
+	return getCGroupsV2PathForContainer(cgroupPath)
+}
+
+// GetPodCgroupsV2AbsPath returns the absolute path to the cgroup v2 directory for a pod sandbox.
+// This method converts relative cgroup paths to absolute paths, for different cgroup managers,
+// QoS classes, and custom cgroup hierarchies.
+// It returns an empty string if the pod has no Linux configuration or cgroups path.
+func GetPodCgroupsV2AbsPath(pod *api.PodSandbox) string {
+	if pod == nil || pod.Linux == nil || pod.Linux.CgroupsPath == "" {
+		return ""
+	}
+
+	cgroupPath := pod.Linux.CgroupsPath
+
+	// For pods, we need to check the runtime type to determine if .scope should be added
+	// NOTE: The podSandbox for the cri-o runtime does not suffix ".scope" to the pod cgroup path
+	runtime := detectContainerRuntime(cgroupPath)
+	isPodScopeRequired := runtime != "crio"
+
+	return getCGroupsV2PathForPodWithRuntime(cgroupPath, isPodScopeRequired)
+}
+
+// Helper functions
+
+// getCGroupsV2PathForContainer helper for container paths
+func getCGroupsV2PathForContainer(cgroupPath string) string {
+	if filepath.IsAbs(cgroupPath) {
+		return cgroupPath
+	}
+
+	cgroupV2Root := getCgroupV2Root()
+	if cgroupV2Root == "" {
+		// Fallback to default cgroup v2 mount point
+		cgroupV2Root = "/sys/fs/cgroup"
+	}
+
+	resolvedPath := resolveCgroupPath(cgroupV2Root, cgroupPath, true)
+	return resolvedPath
+}
+
+// getCGroupsV2PathForPodWithRuntime helper for pod paths with runtime detection
+func getCGroupsV2PathForPodWithRuntime(cgroupPath string, shouldAddScope bool) string {
+	if filepath.IsAbs(cgroupPath) {
+		return cgroupPath
+	}
+
+	cgroupV2Root := getCgroupV2Root()
+	if cgroupV2Root == "" {
+		// Fallback to default cgroup v2 mount point
+		cgroupV2Root = "/sys/fs/cgroup"
+	}
+
+	// Try to resolve the path using different cgroup drivers (cgroupfs, systemd)
+	resolvedPath := resolveCgroupPath(cgroupV2Root, cgroupPath, shouldAddScope)
+	return resolvedPath
+}
+
+// getCgroupV2Root finds the cgroup v2 mount point by reading /proc/mounts
+// It returns an empty string if no cgroup v2 mount point is found
+func getCgroupV2Root() string {
+	commonPaths := []string{
+		"/sys/fs/cgroup",
+		"/cgroup2",
+	}
+
+	if mountPoint := findCgroupV2Mount(); mountPoint != "" {
+		return mountPoint
+	}
+
+	for _, path := range commonPaths {
+		if isCgroupV2Mount(path) {
+			return path
+		}
+	}
+
+	return "/sys/fs/cgroup"
+}
+
+// findCgroupV2Mount reads /proc/mounts to find the cgroup2 filesystem mount point
+func findCgroupV2Mount() string {
+	data, err := os.ReadFile("/proc/mounts")
+	if err != nil {
+		return ""
+	}
+
+	lines := strings.Split(string(data), "\n")
+	for _, line := range lines {
```

Comment:
Wondering if https://pkg.go.dev/github.com/moby/sys/mountinfo should be used; or at least, I know @kolyshkin made various rounds through many projects to get rid of (sometimes _subtly_ wrong) implementations.